### ISSUE TITLE
[bug 🐛] incorrect mamba initialization

### ISSUE THREAD

**Issue by Niccolo-Ajroldi** [athms/mad-lab#3]
## Description

Mamba has several custom layer initializations, which are correctly implemented in this repo.
However, the current version of the model wrapper **overwrites** these initializations.

## More details
`mad/model/layers/mamba.py` uses a custom initialization for the linear layer `dt_proj`:

```python
# Initialize special dt projection to preserve variance at initialization
dt_init_std = self.dt_rank**-0.5 * dt_scale
nn.init.constant_(self.dt_proj.weight, dt_init_std)

# Initialize dt bias so that F.softplus(dt_bias) is between dt_min and dt_max
inv_dt = ...
self.dt_proj.bias.copy_(inv_dt)
self.dt_proj.bias._no_reinit = True
```

But `mad/model/language_model.py` applies Gaussian initialization to all the linear layers, overwriting Mamba initialization:
```python
class LanguageModel(nn.Module):
    def _init_weights(self, m, initializer_range=0.02) -> None:
        if isinstance(m, nn.Linear):
            nn.init.normal_(m.weight, mean=0, std=initializer_range). # overwrites dt_proj.weight
            if m.bias is not None:
                nn.init.zeros_(m.bias). # overwrites dt_proj.bias
```
Moreover, it's now standard practice to initialize only the embedding layers with std=0.002, not all Linear layers. [Mamba implementation](https://github.com/state-spaces/mamba/blob/f0affcf69f06d1d06cef018ff640bf080a11c421/mamba_ssm/models/mixer_seq_simple.py#L63) does it as well.

## Fix
I opened a pull-request https://github.com/athms/mad-lab/pull/4 to reflect the correct initialization of Mamba.

**Comment by bilgeacun** [comment]
This commit is resulting in much longer train times to reach the same threshold accuracy. 
For example with the simple train example:
`python -m train --task in-context-recall --layers mh-attention swiglu mh-attention swiglu`, it requires **13 epochs** to reach > 0.999 accuracy. Whereas after this initialization, it takes **110 epochs to reach the same level**.
